src/rpc/connectivity/cluster.cc

```diff
@@ -1037,27 +1067,27 @@ void connectivity_cluster_t::run_t::handle(
     std::set<host_and_port_t> other_peer_addr_hosts;
     if (deserialize_universal_and_check(conn, &other_id, peername) ||
         deserialize_universal_and_check(conn, &other_peer_addr_hosts, peername)) {
-        return;
+        return join_result_t::TEMPORARY_ERROR;
     }
 
     {
         // Tell the other node that we are happy to connect with it
         write_message_t wm;
         serialize_universal(&wm, handshake_result_t::success());
         if (send_write_message(conn, &wm)) {
-            return; // network error.
+            return join_result_t::TEMPORARY_ERROR; // network error.
         }
 
         // Check if there was an issue with the connection initiation
         handshake_result_t handshake_result;
         if (deserialize_universal_and_check(conn, &handshake_result, peername)) {
-            return;
+            return join_result_t::TEMPORARY_ERROR;
         }
         if (handshake_result.get_code() != handshake_result_code_t::SUCCESS) {
             logWRN("Remote node refused to connect with us, peer: %s, reason: \"%s\"",
                    peername,
                    sanitize_for_logger(handshake_result.get_error_reason()).c_str());
-            return;
+            return join_result_t::PERMANENT_ERROR;
```

Comment:
Assuming there's at least one failed handshake that's a TEMPORARY_ERROR on the sender's side (see above for the admin password case for example), this must also be considered a TEMPORARY_ERROR.

You could distinguish based on the error code in the handshake if you think that it's important to return a PERMANENT_ERROR here for some cases (is it?).